File: 47. LCS Problem/047.py

def test():
    s1 = input()
    s2 = input()
    len1 = len(s1)
    len2 = len(s2)
    table = [0 for i in range(len2+1)]
    ans = 0
    for i in range(len1):
        for j in range(len2-1, -1, -1):
            if s1[i] == s2[j]:
                if i == 0 and j == 0:
                    table[j] = 1
                else:
                    table[j] = table[j-1] + 1
            else:
                table[j] = 0
            ans = max(ans, table[j])
    return ans


# -- author: lijw --
if __name__ == '__main__':
    T = int(input())
    for _ in range(T):
        print(test())

File: 47. LCS Problem/047.cpp
#include<stdio.h>
#include<string.h>
char s1[2005],s2[2005];
int len1,len2;
int main(){
	int t,i,j,k,ans;
	scanf("%d",&t);
	while(t--){
		ans=0;
		scanf("%s %s",&s1,&s2);
		len1=strlen(s1);len2=strlen(s2);
		for(i=0;i<len1;i++){
			if(len1-i<=ans)
			break;
			for(j=0;j<len2;j++){
				if(s1[i]==s2[j]){
					k=1;
					while(s1[i+k]!='\0'&&s2[j+k]!='\0'&&s1[i+k]==s2[j+k])
					k++;
					if(ans<k)
					ans=k;
				}
			}
		}
		printf("%d\n",ans);
	}
	return 0;
}
